src/util/keyboardVisibility.ts

```diff
@@ -0,0 +1,63 @@
+import _ from 'lodash'
+import { editingActionCreator as editingAction } from '../actions/editing'
+import { isTouch } from '../browser'
+import { KEYBOARD_VISIBILITY_THRESHOLD } from '../constants'
+import * as selection from '../device/selection'
+import store from '../stores/app'
+
+let lastViewportHeight: number | null = window.visualViewport?.height || null
+
+/**
+ * Monitors viewport resize events to detect virtual keyboard visibility changes.
+ *
+ * This is primarily needed for Android devices when a user navigates away using
+ * the back button (dismissing the keyboard) or other UI elements, it won't trigger
+ * the blur event. In contrast, iOS typically handles this automatically.
+ *
+ * When a significant viewport height increase is detected (indicating keyboard closing),
+ * this utility dispatches an action to exit editing mode, preventing the UI from
+ * remaining in an inconsistent state where editing UI controls are visible but the
+ * keyboard is closed.
+ 
+ * @returns A throttled function that should be attached to viewport resize events.
+ */
+const handleKeyboardVisibility = _.throttle(() => {
+  // Only apply this for touch devices  where the issue occurs
+  if (!isTouch || !window.visualViewport) {
+    return
+  }
+
+  const visualViewport = window.visualViewport
```

Comment:
Reassigning `window.visualViewport` is unnecessary. Object property dereference is fast, and adding an additional identifier to the scope to save 7 characters is not worth it.